### PR TITLE
Fyle Archive Utility - Expenses PR

### PR DESCRIPTION
**Pull request by Sravanksk**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,37 @@
...
+
+     
+#Usage 
+ 
+To use this archive utility you'll need these Fyle credentials : client ID, client secret and refresh token.
+
+*  Make connection to this archive utility by using the following command
+
+
+       python main.py connect
```
Comment (Shwetabhk) [inline README.md]: please don't use main.py, since this will be a python module released to PyPI
we can do something like 
`python -m fyle_archive_utility <commandline args>`

Also, connect does not have to be a separate command, credentials are not there in the ~/.fylesdkrc then ask for them, otherwise just connect with existing ones.
example
```
$ python -m fyle_archive_utility --download=expenses --format=csv
# Then this should ask
Please enter the credentials:
base_url: <type> press enter
client_id: <type> press enter
client_secret: <type> press enter
refresh_token: <type> press enter

# save this info in ~/.fylesdkrc file
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,37 @@
...
+           
+             
+* Now you are able to access the command line interface for downloading Fyle data
+
+
+       python main.py expenses csv  // Lists out all the expenses of your org and will save it in a csv file
```
Comment (Shwetabhk) [inline README.md]: Please use names for each argument, can check the comment above, also can check ops jobs for this

File: README.md [hunk 3]
```diff
@@ -1,2 +1,47 @@
...
+This project requires python 3.7+ , fylesdk  and click library
+
+* You can download this project and use it ( copy in your own project, etc)
+* Install it from [pip](https://pypi.org)
+
+      $ pip install fyle-archive-utility
+
+
+     
+#Usage
```
Comment (Shwetabhk) [inline README.md]: space between '#' and text, otherwise the readme wont load well

File: README.md [hunk 4]
```diff
@@ -1,2 +1,47 @@
...
+
+       $ python -m fyle_archive_utility --download=expenses --format=csv     // Lists out all the expenses of your org and saves it in a csv file
+       
+       $ python -m fyle_archive_utility --download=expenses --format=json     // Lists out all the expenses of your org and saves it in a json file
+
+       $ python -m fyle_archive_utility --download=expenses_with_attachements     // Downloads the last 31 days expenses of your org with attachments in zip file
+
+       $ python -m fyle_archive_utility --download=expenses --state='COMPLETE' --approved_at='2019-10-31T07:08:01.873Z' --format=csv     // Lists out all the expenses of your org with the given filters and saves it in a csv file
+       
+#License
```
Comment (Shwetabhk) [inline README.md]: space between '#' and text, otherwise the readme wont load well

File: README.md [hunk 5]
```diff
@@ -1,2 +1,47 @@
 # fyle-archive-utility
+
```
Comment (Shwetabhk) [inline README.md]: Please remove all unnecessary lines, atmost leave single line spaces

File: README.md [hunk 6]
```diff
@@ -1,2 +1,47 @@
...
+
+     
+#Usage 
+ 
+To use this archive utility you'll need these Fyle credentials : client ID, client secret and refresh token.
+
+*  Make connection to this archive utility by using the following command
+
+       $ python -m fyle_archive_utility --connect
```
Comment (Shwetabhk) [inline README.md]: Please don't state connect separately, whenever someone pulls up expenses ask them for credentials if rc file is empty otherwise proceed with creds in rc file

File: archive_utility/app.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+@click.option('--clientid', prompt=True, help='Enter Your CLIENT_ID')
+@click.option('--clientsecret', prompt=True, help='Enter Your CLIENT_SECRET')
+@click.option('--refreshtoken', prompt=True, help='Enter Your REFRESH_TOKEN')
+@click.option('--baseurl', prompt=True, help='Enter Your BASE_URL')
+def connect(clientid, clientsecret, refreshtoken, baseurl):
+    """ Make connection to Fyle with credentials """
+    credentials_file = os.path.expanduser('~/.fylesdkrc')
+    with open(credentials_file, 'w') as cfg:
+        cfg.write('{} {} {} {}'.format(clientid, clientsecret, refreshtoken, baseurl))
+        print("Your Credentials are saved")
```
Comment (snarayanank2) [inline archive_utility/app.py]: no print statements

File: archive_utility/app.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+@main.command()
+@click.option('--clientid', prompt=True, help='Enter Your CLIENT_ID')
+@click.option('--clientsecret', prompt=True, help='Enter Your CLIENT_SECRET')
+@click.option('--refreshtoken', prompt=True, help='Enter Your REFRESH_TOKEN')
+@click.option('--baseurl', prompt=True, help='Enter Your BASE_URL')
+def connect(clientid, clientsecret, refreshtoken, baseurl):
+    """ Make connection to Fyle with credentials """
+    credentials_file = os.path.expanduser('~/.fylesdkrc')
+    with open(credentials_file, 'w') as cfg:
+        cfg.write('{} {} {} {}'.format(clientid, clientsecret, refreshtoken, baseurl))
```
Comment (snarayanank2) [inline archive_utility/app.py]: Write this as a json file

File: archive_utility/app.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+
+@main.command()
+@click.option('--format', help="Enter the format of the file 'csv' or 'json' ")
+@click.option('--state', help="Enter the state of Expenses [ 'PAID' , 'APPROVED' ]")
+@click.option('--approved_at')
+def expenses(format, state, approved_at):
+    """ Download the list of expenses with required filters """
+    try:
+        cred_file = os.path.expanduser('~/.fylesdkrc')
```
Comment (snarayanank2) [inline archive_utility/app.py]: separate logic of initializing sdk from config file into a separate function. this will be reused in multiple places

File: README.md [hunk 10]
```diff
@@ -1,2 +1,42 @@
...
+       base_url: <type> press enter
+       client_id: <type> press enter
+       client_secret: <type> press enter
+       refresh_token: <type> press enter
+
+Once you are connected your credentials will be saved and you can access your data 
+                      
+Now you are able to access the command line interface for downloading Fyle data
+
+       $ python -m fyle_archive_utility --download=expenses --format=csv     // Lists out all the expenses of your org and saves it in a csv file
```
[the remark this answers is not in the document]
Comment (snarayanank2) [reply README.md]: this has not been updated yet

File: archive_utility/app.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                                                        approved_at=approved_at,
+                                                        )
+            JSON_DATA.load_data_to_json(response_data)
+        elif format == "csv":
+            response_data = connection.Expenses.get_all(state=state,
+                                                        approved_at=approved_at,
+                                                        )
+            CSV_FORMAT.load_data_to_csv(response_data)
+        # If the file is empty , ask the user to reconnect again
+        elif os.stat(cred_file).st_size == 0:
```
Comment (snarayanank2) [inline archive_utility/app.py]: this should be handled in sdk initialization, not here

File: archive_utility/csv_format.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
+""" Produces data in CSV format """
+
+import csv
+import os
+from datetime import datetime
+import click
+
+
+class CsvHandler:
```
Comment (snarayanank2) [inline archive_utility/csv_format.py]: have a single class called Dumper which has two static methods for writing array of dicts to json or csv file.

```
class Dumper:

   @staticmethod
   def dump_csv(array_dicts, filename):
       pass

   @staticmethod
   def dump_json(array_dicts, filename):
       pass

```

File: archive_utility/app.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    :return: Expenses stored in a file that matches the above filter parameters.
+    """
+    if format == "json":
+        response_data = fyle_connection.extract_expenses(state=state)
+        logger.info('Downloading data from Fyle in JSON format.')
+        dumper.Dumper.dump_json(response_data, path)
+    elif format == "csv":
+        response_data = fyle_connection.extract_expenses(state=state)
+        logger.info('Downloading data from Fyle in CSV format.')
+        dumper.Dumper.dump_csv(response_data, path)
```
Comment (Shwetabhk) [inline archive_utility/app.py]: instead of doing `dumber.Dumper.dump_csv` you can change the wait it is imported
you can do
```python
from archive_utility.dumper import Dumper
# simply call Dumper
Dumper.dump_csv
```

File: archive_utility/app.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+@click.option('--format', help="Enter the format of the file 'csv' or 'json' ")
+@click.option('--state', help="Enter the state of Expenses [ 'PAID' , 'APPROVED' ]")
+@click.option('--path', help="Enter the directory where you want to save your file", required=True)
+def expenses(format, state, path):
+    """
+
+    :param format: format of the file to be generated 'CSV' or 'JSON'
+    :param state:  state of the Expense [ 'PAID' , 'DRAFT' , 'APPROVED' , 'APPROVER_PENDING' , 'COMPLETE' ]
+    :param path:   Takes the path of the file to save the data.
+    :return: Expenses stored in a file that matches the above filter parameters.
```
Comment (Shwetabhk) [inline archive_utility/app.py]: This here means that the function has a return type
Since it does not have a return type can remove this

File: archive_utility/csv_format.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
+""" Produces data in CSV format """
```
Comment (Shwetabhk) [inline archive_utility/csv_format.py]: Is this file needed anymore?

File: archive_utility/dumper.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        """
+        :param data: Takes existing Expenses Data, that match the parameters
+        :param path: Takes the path of the file
+        :return:  JSON file with the list of existing Expenses, that match the parameters
+        """
+        filename = path + "/{0}--Date--{1}.json"
+        os.makedirs(os.path.dirname(filename), exist_ok=True)
+        with open(filename.format(data[0]['org_name'], datetime.now()), 'w') as export_file:
+            logger.info('Downloading data from Fyle in JSON format.')
+            json.dump(data, export_file, indent=4, sort_keys=True)
```
Comment (Shwetabhk) [inline archive_utility/dumper.py]: You can use the same thing for reading from the configuration file

File: archive_utility/json_format.py [hunk 17]
```diff
@@ -0,0 +1,17 @@
+""" Produces data in JSON format """
```
Comment (Shwetabhk) [inline archive_utility/json_format.py]: I think this file can also be removed

File: requirements.txt [hunk 18]
```diff
@@ -0,0 +1,3 @@
+logger==1.4
+Click==7.0
```
Comment (Shwetabhk) [inline requirements.txt]: small c in click

File: setup.py [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+setuptools.setup(
+    name='fyle-archive-utility',
+    version='0.1.0',
+    author='Fyle',
+    author_email='sravan.kumar@fyle.in',
+    description='Command-line tool to download Fyle data.',
+    license='MIT',
+    long_description=long_description,
+    long_description_content_type='text/markdown',
+    include_package_data=True,
```
Comment (Shwetabhk) [inline setup.py]: Can remove this line

File: setup.py [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    keywords=['fyle', 'archive', 'python', 'sdk'],
```
Comment (Shwetabhk) [inline setup.py]: add other relevant keywords like data, download, archive, etc

File: archive_utility/app.py [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+@click.option('--refresh_token', prompt='refresh_token', help='Enter Your REFRESH_TOKEN')
+@click.option('--base_url', prompt='base_url', help='Enter Your BASE_URL')
+def connect(client_id, client_secret, refresh_token, base_url):
+    """
+    :param client_id: Client ID for Fyle API.
+    :param client_secret: Client secret for Fyle API.
+    :param refresh_token: Refresh Token for Fyle API.
+    :param base_url: BaseURL.
+    """
+    try:
```
Comment (Shwetabhk) [inline archive_utility/app.py]: Is try-catch necessary here?

File: archive_utility/app.py [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+    else:
+        logger.warning('Some of the parameters are wrong')
+
+
+if __name__ == "__main__":
+    # Extracting FYLE Credentials from the json file
+    try:
+        path_to_json = os.path.expanduser('~/.config.json')
+        with open(path_to_json, "r") as config_file:
+            my_dict = json.load(config_file)
```
Comment (Shwetabhk) [inline archive_utility/app.py]: dont call it `my_dict` call it something like `credentials` instead

File: archive_utility/app.py [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+            config_file.close()
+        CLIENT_ID, CLIENT_SECRET, REFRESH_TOKEN, BASE_URL = (my_dict["client_id"]), (my_dict["client_secret"]), (
```
Comment (Shwetabhk) [inline archive_utility/app.py]: I think the brackets are not needed
```python
CLIENT_ID, CLIENT_SECRET, REFRESH_TOKEN, BASE_URL = credentials['client_id'], credentials['client_secret'], credentials['refresh_token'], credentials['base_url']
```

File: archive_utility/app.py (hunk 21 above)
Comment (Sravanksk) [reply archive_utility/app.py]: Yes , It'll throw an error message if the file is empty or not existing

Comment (Sravanksk) [reply archive_utility/app.py]: Yes , we can remove this , won't be a problem

**Comment by Shwetabhk** [comment]
Fyle Archive Utility - initial commit - Download of expenses in CSV and JSON
